src/common/metrics/ngx_wa_metrics.c

```diff
@@ -0,0 +1,460 @@
+    metrics->shm_zone->noreuse = 0;
+
+    if (old_metrics
+        && (metrics->workers != old_metrics->workers
+            || metrics->config.slab_size != old_metrics->config.slab_size))
+    {
+        metrics->shm_zone->noreuse = 1;
+    }
+
+    return NGX_CONF_OK;
+}
+
+
+ngx_int_t
+ngx_wa_metrics_init(ngx_wa_metrics_t *metrics, ngx_cycle_t *cycle)
+{
+    ngx_wasm_shm_kv_t  *old_shm_kv;
+
+    if (metrics->old_metrics && !metrics->shm_zone->noreuse) {
+        /* reuse old kv store */
+        metrics->shm->data = metrics->old_metrics->shm->data;
+        return NGX_OK;
+    }
+
+    if (ngx_wasm_shm_kv_init(metrics->shm) != NGX_OK) {
+        return NGX_ERROR;
+    }
+
+    if (metrics->old_metrics && metrics->shm_zone->noreuse) {
+        old_shm_kv = ngx_wasm_shm_get_kv(metrics->old_metrics->shm);
+
+        return realloc_metrics(metrics, old_shm_kv->rbtree.root,
+                               old_shm_kv->rbtree.sentinel);
+    }
+
+    return NGX_OK;
+}
+
+
+ngx_int_t
+ngx_wa_metrics_define(ngx_wa_metrics_t *metrics, ngx_str_t *name,
+    ngx_wa_metric_type_e type, uint32_t *out)
+{
+    ssize_t           size = sizeof(ngx_wa_metric_t)
+                             + sizeof(ngx_wa_metric_val_t) * metrics->workers;
+    uint32_t          cas, mid;
+    ngx_int_t         rc, written;
+    ngx_str_t        *p, val;
+    ngx_wa_metric_t  *m;
+    u_char            buf[size];
+
+    if (type != NGX_WA_METRIC_COUNTER
+        && type != NGX_WA_METRIC_GAUGE
+        && type != NGX_WA_METRIC_HISTOGRAM)
+    {
+        return NGX_ABORT;
+    }
+
+    if (name->len > metrics->config.max_metric_name_length) {
+        return NGX_ABORT;
+    }
+
+    mid = ngx_crc32_long(name->data, name->len);
+
+    ngx_wasm_shm_lock(metrics->shm);
+
+    rc = ngx_wasm_shm_kv_get_locked(metrics->shm, NULL, &mid, &p, &cas);
+    if (rc == NGX_OK) {
+        ngx_log_debug1(NGX_LOG_DEBUG_WASM, metrics->shm->log, 0,
+                       "wasm returning existing metric id \"%uD\"", mid);
+        goto done;
+    }
+
+    ngx_memzero(buf, size);
+    m = (ngx_wa_metric_t *) buf;
+    m->type = type;
+
+    if (type == NGX_WA_METRIC_HISTOGRAM) {
+        rc = ngx_wa_metrics_histogram_add_locked(metrics, m);
+        if (rc != NGX_OK) {
+            goto error;
+        }
+    }
+
+    val.len = size;
+    val.data = buf;
+
+    rc = ngx_wasm_shm_kv_set_locked(metrics->shm, name, &val, 0, &written);
+    if (rc != NGX_OK) {
+        goto error;
+    }
+
+done:
+
+    *out = mid;
+
+error:
+
+    ngx_wasm_shm_unlock(metrics->shm);
+
+    if (rc == NGX_OK) {
+        ngx_wasm_log_error(NGX_LOG_INFO, metrics->shm->log, 0,
+                           "defined %V \"%V\" with id %uD",
+                           ngx_wa_metric_type_name(type), name, mid);
+    }
+
+    return rc;
+}
+
+
+ngx_int_t
+ngx_wa_metrics_increment(ngx_wa_metrics_t *metrics, uint32_t mid, ngx_int_t n)
+{
+    uint32_t          cas;
+    ngx_uint_t        slot;
```

Comment:
nitpick: maybe declare the variable `slot` after `val` so that variable declarations are sorted by the length of the type name.

Reply:
Length of type name isn't really the sorting criteria

Reply:
In this specific case, I just didn't like that a stack value was in between 2 pointer values, so I moved it up along the other unsigned integer variable. Sometimes re-arranging by length is more aesthetic but in this case I think this is more neat.